distributed/scheduler.py

```diff
@@ -5636,3 +5659,21 @@ def remove_worker(self, worker=None, **kwargs):
 
     def teardown(self):
         self.bcomm.close()
+
+
+class CollectTaskMetaDataPlugin(SchedulerPlugin):
+    def __init__(self, scheduler, name):
+        self.scheduler = scheduler
+        self.name = name
+        self.keys = set()
+        self.metadata = {}
+
+    def update_graph(self, scheduler, dsk=None, keys=None, restrictions=None, **kwargs):
+        self.keys.update(keys)
+
+    def transition(self, key, start, finish, *args, **kwargs):
+        if start == "processing" and (finish == "memory" or finish == "erred"):
+            ts = self.scheduler.tasks.get(key)
+            if ts is not None and ts.key in self.keys:
+                self.metadata[key] = ts.metadata
```

Comment:
Are there other parts of the `TaskState` state that would be valuable to include?

Reply:
That's a good point. The `finish` state would also be useful to have.